docs/messaging/messaging-metrics.md

```diff
@@ -32,6 +30,7 @@ All messaging metrics share the same set of attributes:
 <!-- semconv metric.messaging.attributes(full) -->
 | Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
 |---|---|---|---|---|---|
+| [`messaging.operation.name`](/docs/attributes-registry/messaging.md) | string | The system-specific name of the messaging operation. | `ack`; `nack`; `send` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

Comment:
I can still publish messages using different operations like `schedule` or `send` within a single system. It also captures system-specific terminology, so does not seem to hurt.

Reply:
Should we go a step further and make the counters as generic as the duration metrics? So we'd have two symmetric pairs of metrics:
* `messaging.client.operation.duration`
* `messaging.client.operation.messages`
* `messaging.client.process.duration`
* `messaging.client.process.messages`

This has the advantage of providing proper counters for _all_ operations. On the other hand, it is a bit more cumbersome to use (one needs to know the system-specific operation name to be able to count produced messages).

> so does not seem to hurt.

I agree it doesn't hurt and this isn't a blocker, I just think that for most cases this (required) attribute doesn't provide much use on counter metrics.

Reply:
As discussed in the workgroup meeting, we decided that we will further explore the generic metrics proposed above (exact names are still open for discussion):
* `messaging.client.operation.duration`
* `messaging.client.operation.messages`
* `messaging.client.process.duration`
* `messaging.client.process.messages`

This will also require the presence of the `messaging.operation.type` attribute, so that generic operation counts can be aggregated towards standardized operation types.

### Example

In this example two messages are published (one via a "send" operation, the other one via a "schedule" operation), for one message also a "create" operation was instrumented:
```
messaging.client.operation.messages{messaging.operation.type=create} 1
messaging.client.operation.messages{messaging.operation.type=publish, messaging.operation.name=send} 1
messaging.client.operation.messages{messaging.operation.type=publish, messaging.operation.name=schedule} 1
```
Here a count of all operations with the type "publish" would give the number of all published messages (initially the dedicated `messaging.client.published.message` was intended for this).

Then two messages are received and both are settled via one "commit" operation:

```
messaging.client.operation.messages{messaging.operation.type=receive, messaging.operation.name=consume} 2
messaging.client.operation.messages{messaging.operation.type=settle, messaging.operation.name=commit} 1
```

Here a count of all operations with the type "receive" would give the number of all consumed messages (initially the dedicated `messaging.client.consumed.message` was intended for this).

### Problem

It's not clear to me how to reliable count consumed messages with this approach. There could possibly be just a single "receive" operation for a message (pull-based consumption with out "process" instrumentation), there could be just a single "process" operation for a message (push-based consumption without "receive" instrumentation), or there could be both "receive" and "process" operations for a single message (pull-based consumption with "process" instrumentation).

Given this situation, it's not clear to me how to deduce the number of consumed messages from `messaging.client.operation.messages` and `messaging.client.process.messages`.

Reply:
Given that the number of consumed messages can't be deduced from the number of processed and the number of received messages, I propose to keep generic operation metrics for flexibility, but to additionally add counters for produced and consumed messages:

* `messaging.client.operation.duration`
* `messaging.client.operation.messages`
* `messaging.client.process.duration`
* `messaging.client.process.messages`
* `messaging.client.produced.messages`
* `messaging.client.consumed.messages`

Reasons:
* We agreed that "process" is a special operation that's important for default dashboards, so we keep this as a separate metric.
* All other operations are tracked via the generic "operation" metrics (similar to database metrics).
* We keep track of counts of produced and consumed messages independent of operations. This might create some redundancy, as the count of produced messages could be deduced from operation counts, given we keep `messaging.operation.type`. However, keeping dedicated counters here makes the whole model more clear and user-friendly, and also enables us to go ahead without `messaging.operation.type`.

@lmolkova @joaopgrassi @trask  What do you think?

Reply:
Also: In pull based scenarios, we will always have a "receive" operation, so for that case we can consistently count the received messages. The problem seems to be on push-based scenarios, as there we can't guarantee that a process span/operation will happen as it's in user land code.

Reply:
it feels like we're trying to make metrics work with existing operation types. What if we change operation types instead?
what if we introduced a `consume` operation type instead of `process`. `process` is not a messaging operation type - it's special. We need the terminology, but not necessarily a type

Let's do a mental experiment:

- `consume` operation captures that a message is dispatched to the application
   - it's pretty much the same as `delivery` operation, just a different word since `delivered.messages` is ambiguous
   - message is consumed when it's dispatched to the application, it's counted once either at the end of `receive` or right before `process` starts
  - we need to document this moment anyway to tell when to update the message counter and to provide time-in-queue/delay metrics
- spans: 
    - push-based: `consume` type should be set on the process span. 
    - pull-based: no `consume` span
- metrics: 
  - `messaging.client.operation.duration` is never reported for `consume` operation type - it's in the `process.duration`
  - `messaging.client.operation.messages` is reported for it 

So we'd have:
- `messaging.client.operation.duration`: for `publish`, `receive` and untyped operations
- `messaging.client.operation.messages`: for all defined operation types and untyped operations except processing
- `messaging.client.process.duration`: for processing duration
- `messaging.client.process.messages`: for processing throughput

WDYT?

Reply:
it feels like we're trying to make metrics work with existing operation types. What if we change operation types instead?
what if we used `receive` for all kinds of deliveries (push and pull). Maybe we can come up with a better name for it?

Let's do a mental experiment:

- `receive` operation captures a process or receiving a message in one way or another. 
- spans: 
    - `receive` is always created. It's always a consumer span.
    - `process` can be created in addition to `receive`
- metrics: 
  - `messaging.client.operation.duration` is not reported for  push-based `receive` operation type - it's in the `process.duration`. 
  - `messaging.client.operation.messages` is reported for all operations that deal with messages

So we'd have:
- `messaging.client.operation.duration`: for all operations except push-based `receive` and `process`
- `messaging.client.operation.messages`: for all operations except processing
- `messaging.client.process.duration`: for processing duration
- `messaging.client.process.messages`: for processing throughput

TBH I no longer understand why we want to differentiate process operation. 

WDYT?

Reply:
Update the PR with this revolutionary suggestion to demonstrate it. Let's discuss on the sig meeting.

Reply:
Thanks for your feedback and ideas.

> what if we used receive for all kinds of deliveries (push and pull). Maybe we can come up with a better name for it?

Wouldn't that make it meaningless to compare the duration of "receive" operations, when it could measure either a push or a pull operation? Also, would this mean that one needs two spans to capture "process" operations (as "receive" will always be required)?

> it feels like we're trying to make metrics work with existing operation types.

From my point of view that's not a strict requirement. At a minimum we need to be able to align duration metrics with span durations.

We should strive to design metrics so that the following questions can be easily answered, ideally without knowing many details about system-specifics:
* How many message were produced?
* How many messages were consumed?
* What's the processing duration?
* What's the error rate for any of the operations above (producing, consuming, and processing)?

For other questions I wouldn't mind if it would be a bit more complicated (and maybe involve system-specific knowledge) to extract information from metrics. Examples include receive duration, publish duration, or settlement related metrics.

Reply:
Based on the discussion I changed the PR to 
- keep operation.type
- no changes around operation types
- introduce following metrics:
  - messaging.client.request.duration (for all operations except processing)
  - messaging.process.duration 
  - messaging.client.published.messages 
  - messaging.client.consumed.messages (all consumed messages, either received or processed, counted once per message delivery)

As a result:

Spans: 
- consumer span represents either receive or process operation depending on the scenario. `msg.operation.type=process` tells it's processing. It can be used to convert spans to metrics.

Metrics:
- Message throughput: `messaging.client.published.messages` for produced, `messaging.client.consumed.messages` for consumed. No metric for count of processed messages. We can always add one later.
- Request throughput is derived from `messaging.client.request.duration` grouped by operation name/type and `messaging.process.duration`
- Error rate: derived from `messaging.client.request.duration`  and `messaging.process.duration`


Naming and other suggestions are very welcome.